Add FilePosition in FileSelector and ClasspathResourceSelector

timtebeek opened the pull request:
## Overview

Fixes #2146.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

timtebeek commented on junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/FilePosition.java:
```diff
@@ -0,0 +1,178 @@
+	private final int line;
+	private final Integer column;
+
+	protected FilePosition(int line) {
+		Preconditions.condition(line > 0, "line number must be greater than zero");
+		this.line = line;
+		this.column = null;
+	}
+
+	protected FilePosition(int line, int column) {
```
Wasn't particularly pleased about having to make these constructors `protected` for use from `org.junit.platform.engine.support.descriptor.FilePosition` constructors, but saw no way around it. Is there any way to discourage usage through `@API` annotations?

timtebeek commented on platform-tests/src/test/java/org/junit/platform/engine/discovery/FilePositionTests.java:
```diff
@@ -31,12 +32,7 @@
  * @since 1.0
  */
 @DisplayName("FilePosition unit tests")
-class FilePositionTests extends AbstractTestSourceTests {
-
-	@Override
-	Stream<FilePosition> createSerializableInstances() {
-		return Stream.of(FilePosition.from(42, 99));
-	}
```
`AbstractTestSourceTests` is only accessible from `org.junit.platform.engine.support.descriptor`, so I though it preferable to keep that that way. Should this particular override and the functionality it covered be reproduced here? Or should I not yet have moved this tests class at this time?

timtebeek commented on junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/FilePosition.java:
```diff
@@ -0,0 +1,178 @@
+ *
+ * @since 1.0
+ */
+@API(status = STABLE, since = "1.0")
```
Not particularly sure about the `since` argument here, and/or whether to declare this as `STABLE` already, given that we likely want to change the constructors from `protected` to `private`.

mpkorstanje replied:
You can also apply the annotation to specific methods to mark them as internal. The most specific annotation applies.

mpkorstanje commented on junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/FilePosition.java:
```diff
@@ -10,32 +10,35 @@
-@API(status = STABLE, since = "1.0")
-public class FilePosition implements Serializable {
+@API(status = Status.DEPRECATED, since = "5.7")
+public class FilePosition extends org.junit.platform.engine.discovery.FilePosition {
```
IDEs won't pick up on the @API annotation. Adding `@Deprecated` might be useful here.

mpkorstanje commented on junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/FilePosition.java:
```diff
@@ -10,32 +10,35 @@
+@API(status = Status.DEPRECATED, since = "5.7")
+public class FilePosition extends org.junit.platform.engine.discovery.FilePosition {
 
 	private static final long serialVersionUID = 1L;
```
Where are file positions de-serialized too?

marcphilipp replied:
We don't use the serialization ourselves. It's meant for third-party tools that send test events over JVM boundaries. Since we didn't change any fields, I think we should be fine.

mpkorstanje commented on junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/FilePosition.java:
```diff
@@ -77,7 +80,7 @@ public static FilePosition from(int line, int column) {
 	 * @see #from(int, int)
 	 */
-	public static Optional<FilePosition> fromQuery(String query) {
+	public static Optional<org.junit.platform.engine.discovery.FilePosition> fromQuery(String query) {
```
Returning the super class here will break existing callers.

timtebeek replied:
This is slightly embarrassing, but when I change it to `FilePosition` it refuses to compile:
```
> Task :junit-platform-engine:compileJava FAILED
/junit5/junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/FilePosition.java:85: error: fromQuery(String) in org.junit.platform.engine.support.descriptor.FilePosition cannot hide fromQuery(String) in org.junit.platform.engine.discovery.FilePosition
        public static Optional<FilePosition> fromQuery(String query) {
                                             ^
  return type Optional<org.junit.platform.engine.support.descriptor.FilePosition> is not compatible with Optional<org.junit.platform.engine.discovery.FilePosition>
1 error
``` 
I fail to understand why this is only an issue for this static method and not the other two static `from` methods that take one or more `int` arguments. Any suggestions on what to do here?

timtebeek replied:
I've briefly explored using `Optional<? extends FilePosition>` in the superclass `FilePosition`, but that leaks out into usage of that method regrettably. :\

mpkorstanje replied:
A method can hide a method in the parent class if it follows he Liskov Substitution Principle. So a subclass must return a more specific type then the parent class. For generics this inversed.

Suppose it would be possible to do extend a class and return a more specialized generic:

```java
class Parent {
    public List<Number> funX() {
        return new ArrayList<>();
    }
}

class Child extends Parent {
    public List<Double> funX() {
        return new SpecializedDoubleList<>();
    }
}
```

Then by invoking `funX` on the parent type we could get a list that is only intended to contain `Double` values and add an `Integer` to it.

```java
Parent p = new Child();
List<Number> numbers = p.funX();
numbers.add(Integer.valueOf(1));
```
So the problem is that generic type is too specific. It specifies both the elements that can be added to the list (as a consumer) as well as those that can be taken from it (as a producer). By changing the parent to `extends Number` we can specify that the list contains an unknown type of element, but if you were to taken elements from it, it would at-least extend `Number`.

```java
class Parent {
    public List<? extends Number> funX() {
        return new ArrayList<>();
    }
}
```

mpkorstanje replied:
> I've briefly explored using Optional<? extends FilePosition> in the superclass FilePosition, but that leaks out into usage of that method regrettably. :\

It isn't pretty but I think that is okay. 

It is possible to assign `List<Number>` to `List<? extends Number>` so once the old `FilePosition` and the `extends Number` are removed it shouldn't break anything.

timtebeek replied:
Applied in https://github.com/junit-team/junit5/pull/2253/commits/decb66931b828437d53497fcf091ba3209f36774 as mentioned above; probably best for @marcphilipp to have another close look :)

mpkorstanje commented on junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/ClasspathResourceSelector.java:
```diff
@@ -59,6 +65,14 @@ public String getClasspathResourceName() {
+	 * Get the selected position within the classpath resource as a
+	 * {@link FilePosition}.
+	 */
+	public FilePosition getPosition() {
```
JUnit 5 seems to follow the pattern that nullable return values are always wrapped in an Optional.

https://github.com/junit-team/junit5/blob/3f7fed61f2edcee2a5551dbde75382b1dbe21267/junit-platform-console/src/main/java/org/junit/platform/console/tasks/TreeNode.java#L74-L76

timtebeek commented on junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java:
```diff
@@ -135,6 +135,52 @@ public static FileSelector selectFile(File file) {
+	 * @see #selectDirectory(String)
+	 * @see #selectDirectory(File)
+	 */
+	public static FileSelector selectFile(String path, int line, int column) {
```
Had originally not wanted to go with `selectFile(path, FilePosition.from(12, 34))` as to me it seemed slightly odd to ask the caller to invoke two methods to construct a selector. But I can perfectly live with the change as suggested, and have applied it already.

timtebeek commented:
Ok I've hopefully addressed all of your suggestions as intended; I hope it's ok I marked them as "resolved"; If I shouldn't have done that please let me know!

The only outstanding issue (I think) is this slight gaffe on my part where I need some direction: https://github.com/junit-team/junit5/pull/2253/files#r406852337

marcphilipp commented on junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/ClasspathResourceSource.java:
```diff
@@ -21,6 +21,7 @@
 import org.junit.platform.commons.util.Preconditions;
 import org.junit.platform.commons.util.ToStringBuilder;
 import org.junit.platform.engine.TestSource;
+import org.junit.platform.engine.discovery.FilePosition;
```
This is a breaking change (same above). I'm afraid we'll have to leave the existing `getPosition()` method in place, deprecate it, and add a new `getFilePosition` method that returns an optional of the new type.

timtebeek replied:
Good that you caught that one! I've hopefully make the correct changes in https://github.com/junit-team/junit5/pull/2253/commits/a768cf46290e40cc2cd17b2baee630667b2e3b2e ; Could you have a look if this was what you had intended?

timtebeek replied:
Also deprecated the old methods in: https://github.com/junit-team/junit5/pull/2253/commits/dd3d8f7a41064a1daadad7b1da9f07b8d67f5170

marcphilipp replied:
Thanks for the quick turnaround! I'm afraid there's more, though. The old factory methods in the `Source` classes need to be preserved and deprecated, too.

timtebeek replied:
Sure thing: https://github.com/junit-team/junit5/pull/2253/commits/f80fcb92f6b60b4400548f2723e0839e0fe145ea

timtebeek replied:
Hmm; maybe I was too quick with my change; should this apply to `ClassSource` as well?

timtebeek replied:
Figured `ClassSource` probably needed a restore as well, so covered in https://github.com/junit-team/junit5/pull/2253/commits/147ea9d10ebc4b9e6fc9504d230b104c4bd9c70c

timtebeek commented:
@marcphilipp what would be needed to finish this PR? I'm guessing at least a change to the changelog; anything else?

marcphilipp commented:
@timtebeek I think this looks good! I'll discuss our decision wrt. to the package with the team and will get back to you afterwards.

timtebeek commented:
Just noticed https://github.com/junit-team/junit5/commit/19785134f95ba100346eb2d8b0c24f707535ebe4 in master; with this PR added to the [5.7 M2](https://github.com/junit-team/junit5/milestone/49) milestone, should I merge in master and add a rough outline for the changes contained in this PR?

I know this isn't a high priority new feature or fix, but would nonetheless be nice to see this finished if possible, before any other branches introduce conflicts. :)

marcphilipp commented:
@timtebeek Sorry for the delay. We discussed it briefly in the last team call. The current solution still contains breaking changes so we were a bit torn whether to just duplicate the `FilePosition` class to avoid the messy (not your fault!) deprecation work and the confusion caused by it. Would you mind changing this PR so the `FilePosition` class and its tests are copied and the deprecation is reverted? Sorry for the back ans forth! As for the release notes: yes, please rebase on `master` and add a bullet point.

timtebeek commented:
@marcphilipp No problem at all; I've reverted the deprecation changes, created a clean copy of FilePosition and updated the release notes. Could you have another look? :)

timtebeek commented:
@marcphilipp any update on this issue? Just trying to keep this flowing through; if there's any changes I have to make still let me know!

timtebeek commented:
@juliette-derancourt Thanks a lot for the detailed feedback; I've applied all your suggestions. Let me know if there's anything else! :)

timtebeek commented on junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java:
```diff
@@ -135,6 +135,50 @@ public static FileSelector selectFile(File file) {
+	 * @param path the path to the file to select; never {@code null} or blank
+	 * @param position the position inside the file; may be {@code null}
+	 * @see FileSelector
+	 * @see #selectFile(File, int, int)
```
I've specifically added references to the methods that take one fewer/extra argument, or the same amount of arguments with a single argument of a different type. Hope that's what you intended as well! :)

timtebeek commented on platform-tests/src/test/java/org/junit/platform/engine/discovery/DiscoverySelectorsTests.java:
```diff
@@ -153,6 +186,30 @@ void selectClasspathResources() {
 		// standard use case
 		selector = selectClasspathResource("A/B/C/spec.json");
 		assertEquals("A/B/C/spec.json", selector.getClasspathResourceName());
+
+		// with file position
+		selector = selectClasspathResource("A/B/C/spec.json", FilePosition.from(12, 34));
+		assertEquals("A/B/C/spec.json", selector.getClasspathResourceName());
+		assertEquals(FilePosition.from(12, 34), selector.getPosition().get());
```
```suggestion
```

juliette-derancourt commented:
Thanks for the PR @timtebeek!
